src/main/java/org/wildfly/security/auth/realm/CacheableSecurityRealm.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.auth.realm;
+
+import static org.wildfly.common.Assert.checkNotNullParam;
+
+import org.wildfly.security._private.ElytronMessages;
+import org.wildfly.security.auth.server.IdentityLocator;
+import org.wildfly.security.auth.server.RealmIdentity;
+import org.wildfly.security.auth.server.RealmUnavailableException;
+import org.wildfly.security.auth.server.SecurityRealm;
+import org.wildfly.security.auth.server.SupportLevel;
+import org.wildfly.security.auth.server.event.RealmEvent;
+import org.wildfly.security.cache.RealmIdentityCache;
+import org.wildfly.security.credential.Credential;
+import org.wildfly.security.evidence.Evidence;
+
+/**
+ * A realm which provides caching for {@link RealmIdentity} instances obtained from another realm.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class CacheableSecurityRealm<R extends SecurityRealm> implements SecurityRealm {
+
+    private final R realm;
+    private final RealmIdentityCache cache;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param realm the {@link SecurityRealm} whose {@link RealmIdentity} should be cached..
+     * @param cache the {@link RealmIdentityCache} instance
+     */
+    public CacheableSecurityRealm(R realm, RealmIdentityCache cache) {
+        this.realm = checkNotNullParam("realm", realm);
+        this.cache = checkNotNullParam("cache", cache);
+    }
+
+    @Override
+    public RealmIdentity getRealmIdentity(IdentityLocator locator) throws RealmUnavailableException {
+        return cache.computeIfAbsent(locator, locator1 -> {
```

Comment:
So essentially we are acquiring a RealmIdentity here, and putting it in cache, and letting some unspecified external party determine when the identity should be removed from the cache.

I don't think this is good because it will be difficult or even impossible for users to get this right without the cooperation of the backing realm, which has no knowledge that its identities are being cached.  I think this exposes a huge risk of sporadically allowing authentication with outdated credentials.

I think maybe what we need is for realms to be truly cache aware - *this* is where the "CacheableRealm" name would become correct - wherein you can get a callback which is called when a realm identity is updated in the backing realm, ensuring that it stays current in the cache.